src/BlockDiagonals.jl

```diff
@@ -0,0 +1,241 @@
+module BlockDiagonals
+
+using BlockArrays
+using BlockArrays: AbstractBlockSizes, BlockSizes
+using FillArrays
+using LinearAlgebra
+
+export BlockDiagonal, blocks
+
+"""
+    BlockDiagonal{T, V<:AbstractMatrix{T}} <: AbstractBlockMatrix{T}
+
+A BlockMatrix with square blocks on the diagonal, and zeros off the diagonal.
+"""
+struct BlockDiagonal{T, S<:AbstractBlockSizes} <: AbstractBlockMatrix{T}
+    blocks::Vector{<:AbstractMatrix{T}}
+    block_sizes::S
+    function BlockDiagonal(blocks::Vector{<:AbstractMatrix{T}}, sizes::S) where {T, S<:AbstractBlockSizes}
+        all(is_square, blocks) || throw(ArgumentError("All blocks must be square."))
+        # mapreduce would give Array of Tuples; want Array of Arrays
+        return new{T, S}(blocks, sizes)
+    end
+end
+
+function BlockDiagonal(blocks::AbstractVector{<:AbstractMatrix})
+    # mapreduce would give Array of Tuples; want Array of Arrays
+    bs = BlockSizes([[first(size(b)), last(size(b))] for b in blocks]...)
+    return BlockDiagonal(blocks, bs)
+end
+
+is_square(A::AbstractMatrix) = size(A, 1) === size(A, 2)
+blocks(b::BlockDiagonal) = b.blocks
+
+# AbstractBlockMatrix interface
+BlockArrays.blocksizes(b::BlockDiagonal) = b.block_sizes
+function BlockArrays.blocksize(b::BlockDiagonal, i::Int, j::Int)
+    return first(size(blocks(b)[i])), last(size(blocks(b)[j]))
+end
+BlockArrays.nblocks(b::BlockDiagonal) = length(blocks(b)), length(blocks(b))
+BlockArrays.nblocks(b::BlockDiagonal, dims::Integer) = length(blocks(b))
+function BlockArrays.getblock(b::BlockDiagonal{T}, i::Int, j::Int) where T
```

Comment:
```suggestion
function BlockArrays.getblock(b::BlockDiagonal{T}, i::Integer, j::Integer) where T
```

This isn't type stable, but I am Ok with that, it will be a small union.

Reply:
i think `getblock` might be broken right now anyway...